### PR TITLE
Feature/error messages as string views

### PR DESCRIPTION
**Pull request by goedderz**
### Scope & Purpose

As the length of error messages (as returned by `TRI_errno_string`) is known at compile time, it makes sense to save and return them as `std::string_view`s instead of `char const*`.

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [X] :fire: Performance improvement
- [X] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

Enterprise companion PR: https://github.com/arangodb/enterprise/pull/641

#### Backports:

- [X] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

### Testing & Verification

*(Please pick either of the following options)*

- [X] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [X] This change is already covered by existing tests, such as *scripts/unittest*.
- [ ] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [ ] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

### CODE REVIEW

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13782/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13790/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13810/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13836/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13841/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13856/

**Comment by goedderz** [comment]
image

File: arangod/Aql/QueryWarnings.cpp [hunk 1]
```diff
@@ -53,6 +53,10 @@ void QueryWarnings::registerWarning(int code, std::string const& details) {
   registerWarning(code, details.c_str());
 }
 
+void QueryWarnings::registerWarning(int code, std::string_view details) {
+  registerWarning(code, details.data());
```
Comment (maierlars) [inline arangod/Aql/QueryWarnings.cpp:57]: I don't know if `registerWarning` expects the string to be null-terminated but `data` from a string is not necessarily null-terminated.

File: arangod/Aql/Parser.cpp [hunk 2]
```diff
@@ -176,7 +181,7 @@ void Parser::registerParseError(int errorCode, char const* data, int line, int c
 }
 
 /// @brief register a warning
-void Parser::registerWarning(int errorCode, char const* data, int line, int column) {
+void Parser::registerWarning(int errorCode, std::string_view data, int line, int column) {
   // ignore line and column for now
   _query.warnings().registerWarning(errorCode, data);
```
Comment (maierlars) [inline arangod/Aql/Parser.cpp:186]: is `data` null-terminated?

File: arangod/Aql/Parser.cpp [hunk 3]
```diff
@@ -144,6 +144,11 @@ void Parser::registerParseError(int errorCode, char const* format,
   return registerParseError(errorCode, buffer, line, column);
 }
 
+void Parser::registerParseError(int errorCode, std::string_view format,
+                                char const* data, int line, int column) {
+  return registerParseError(errorCode, format.data(), data, line, column);
```
Comment (maierlars) [inline arangod/Aql/Parser.cpp:149]: `data()` is not null-terminated in general.

**Review by maierlars — CHANGES_REQUESTED** [review]
`std::string_view::data()` does [not](https://en.cppreference.com/w/cpp/string/basic_string_view/data#Notes) return a null-terminated string.

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13894/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13932/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13936/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13937/

**Review by maierlars — APPROVED** [review]

**Comment by goedderz** [comment]
image

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13956/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13957/

**Comment by goedderz** [comment]
http://jenkins.arangodb.biz:8080/job/arangodb-matrix-pr/13958/

**Review by maierlars — APPROVED** [review]

**Pull request merged**